Allow update of drawing items

slvrfn opened the pull request:
A user can now have access to the drawing items, and can provide a short list of items that need to be re-calculated. Intended to be used to replace shapes in-place in the items list. 

One example is re-coloring only certain drawing items, all items do not need to be re-added/re-calculated.

Moved transformColor to a static context that can be used outside the library when the user is creating items.

slvrfn committed:
added facial culling  using the shoelace formula
https://www.cs.uic.edu/~jbell/CourseNotes/ComputerGraphics/VisibleSurfaces.html

https://en.wikipedia.org/wiki/Back-face_culling

https://en.wikipedia.org/wiki/Shoelace_formula

slvrfn committed:
only perform measure operations when necessary

slvrfn committed:
better to remove items than to ignore drawing them

if an item is ignored, it still goes through sorting. If the item is removed all together, don't need to worry about if the item is drawn or not later in processing

slvrfn committed:
dont consider item if it is out of bounds

slvrfn committed:
bug

slvrfn committed:
important typo

slvrfn committed:
used incorrect variable

slvrfn committed:
Update versions

slvrfn committed:
Merge branch 'drawing_improvements'

slvrfn committed:
pull request feedback

slvrfn committed:
add user settable variable for bounds checking of items. 

fixed comments for new setters

slvrfn committed:
Merge remote-tracking branch 'FabianIsometric/master'

slvrfn committed:
added equals and gethashcode to path and point objects

used for comparisons

slvrfn committed:
added getpath to Isometric.Item to make it possible to check which shape this item was apart of

slvrfn committed:
added ability to track shapes by identifier

slvrfn committed:
allow items to be iterated backwards for touch lookup

slvrfn committed:
more efficient to use iterator

slvrfn committed:
added ability to do click lookup with touch region (circle) instead of a singular point

slvrfn committed:
cleanup comments

removed bug with calculating twice

slvrfn committed:
pr feedback

slvrfn committed:
credit where credit is due

slvrfn committed:
Revert "added ability to track shapes by identifier"

This reverts commit 45b6998849c9c52a1c659efdde472447382e454f.

slvrfn committed:
drawing items optionally store shape that they belong to

slvrfn committed:
Merge remote-tracking branch 'FabianIsometric/master'

slvrfn committed:
moved transformcolor to be static in the Color class

slvrfn committed:
add ability to update only certain items to be drawn on user request. Intended to be used to replace shapes in-place in the items list. 

One example is re-coloring only certain drawing items, all items do not need to be re-added/re-calculated

FabianTerhorst commented on lib/src/main/java/io/fabianterhorst/isometric/IsometricView.java:
```diff
@@ -72,6 +74,17 @@ public void setTouchRadius(double touchRadius) {
     }
 
+    //allow user to set current items with the option to re-measure again or not
+    public void updateItems(List<Isometric.Item> items) {
```
The user has the trigger the redrawing of the view manually after this?

slvrfn replied:
i will add call to invalidate()

FabianTerhorst commented on lib/src/main/java/io/fabianterhorst/isometric/Isometric.java:
```diff
@@ -390,7 +369,7 @@ public Item findItemForPosition(Point position, boolean reverseSort, boolean tou
         android.graphics.Path drawPath;
 
-        Item(Item item) {
+        public Item(Item item) {
```
Maybe add an static creating method that makes it easier for the user to create an item without manually using the transformColor method.

FabianTerhorst commented on lib/src/main/java/io/fabianterhorst/isometric/Isometric.java:
```diff
@@ -150,6 +112,23 @@ public void measure(int width, int height, boolean sort, boolean cull, boolean b
+    }
+
+    //allow user to recalculate only certain items
+    public void updateItems(List<Item> items, boolean cull, boolean boundsCheck) {
```
Aren't the interception checks for the none updated items missing when you update an item that is behind an none updated one?

slvrfn replied:
The user would be using this method to directly manipulate the items list returned from `getCurrentItems()`. I will add a comment that they are taking potentially dangerous actions when they use this method (manipulating an item that is covered by another item), but it is still a useful method to have. 

In my own implementation, I get the current items `List lis = getCurrentItems()` and find references to some `Shape s` I care about. I then create new `Item` into `List modItems` and use `updateItems(modItems, true, true)` to perform all of the path calculations with culling and bounds detection which removes the bad faces from the provided list. I then directly remove/replace the items `updateItemsList(lis, modItems)`

FabianTerhorst commented:
Im going to merge #16 first. Do you want to take a look to see how much the performance improved?Mostly just the amount of allocations is reduced. ```isPointInPoly``` is now the slowest calculation when having sort activated.

slvrfn commented:
I definitely believe that about `isPointInPoly`. There can be a whole lot of edges to check when drawing. Do you have any ideas to improve that method?

I'll definitely take a look at the other performance improvements!

slvrfn committed:
comments

slvrfn committed:
static item creation/copy

slvrfn committed:
pr feedback

slvrfn committed:
Merge remote-tracking branch 'origin/master'

slvrfn committed:
initial capacity for list on item lookup

slvrfn committed:
typo

slvrfn committed:
get the drawing item associated with an x/y position

slvrfn committed:
very close but not the cake

slvrfn committed:
important typo and comments

slvrfn committed:
comments on z

slvrfn committed:
reverted unnecessary change

slvrfn committed:
created a field in Point to better represent the usage of the z added in translateIsoToViewPoint

slvrfn committed:
Merge branch 'view_to_iso'

slvrfn commented:
I have added some commits which include the ability to convert to/from Isometric and View coordinate systems.

slvrfn committed:
fixed null bug

slvrfn committed:
access

slvrfn committed:
vectors

slvrfn committed:
change access for children

slvrfn committed:
allow child to define setting/getting the items list

slvrfn committed:
allow child to provide alternate isometric instance

slvrfn committed:
access
